File: preview/shoot.py
"""Screenshot helper (Playwright via system Edge).
  python shoot.py page <index.html> <out.png> <width> <height> [full|viewport] [mobile]
  python shoot.py banners <banners.html> <outdir>
"""
import sys
from pathlib import Path
from playwright.sync_api import sync_playwright

mode = sys.argv[1]
with sync_playwright() as p:
    b = p.chromium.launch(channel="msedge")
    if mode == "page":
        src, out, w, h = Path(sys.argv[2]), Path(sys.argv[3]), int(sys.argv[4]), int(sys.argv[5])
        full = (len(sys.argv) > 6 and sys.argv[6] == "full")
        mobile = (len(sys.argv) > 7 and sys.argv[7] == "mobile")
        ctx = b.new_context(viewport={"width": w, "height": h}, device_scale_factor=2 if mobile else 1,
                            is_mobile=mobile, has_touch=mobile)
        pg = ctx.new_page()
        errs = []
        pg.on("pageerror", lambda e: errs.append(str(e)))
        pg.on("console", lambda m: errs.append("console." + m.type + ": " + m.text) if m.type in ("error", "warning") else None)
        pg.goto(src.resolve().as_uri())
        pg.evaluate("document.fonts.ready")
        pg.wait_for_timeout(3600)  # let the receipt finish printing
        # force any pending reveals so the full-page capture is complete
        pg.evaluate("document.querySelectorAll('.reveal-pending').forEach(e=>e.classList.add('in'))")
        pg.wait_for_timeout(700)
        report = pg.evaluate("""() => ({
          fonts: [...document.fonts].filter(f=>f.status==='loaded').map(f=>f.family).filter((v,i,a)=>a.indexOf(v)===i),
          overflow: document.documentElement.scrollWidth > document.documentElement.clientWidth,
          scrollW: document.documentElement.scrollWidth, clientW: document.documentElement.clientWidth,
          h: document.documentElement.scrollHeight
        })""")
        print("fonts:", report["fonts"]); print("overflow:", report["overflow"], report["scrollW"], report["clientW"], "height:", report["h"])
        print("errors:", errs or "none")
        pg.screenshot(path=str(out), full_page=full)
        print("wrote", out)
    elif mode == "banners":
        src, outdir = Path(sys.argv[2]), Path(sys.argv[3])
        outdir.mkdir(parents=True, exist_ok=True)
        ctx = b.new_context(viewport={"width": 2600, "height": 2000}, device_scale_factor=1)
        pg = ctx.new_page()
        errs = []
        pg.on("pageerror", lambda e: errs.append(str(e)))
        pg.goto(src.resolve().as_uri())
        pg.evaluate("document.fonts.ready")
        pg.wait_for_timeout(1200)
        for el in pg.query_selector_all(".canvas"):
            name = el.get_attribute("data-name")
            box = el.bounding_box()
            el.screenshot(path=str(outdir / f"{name}.png"))
            print("wrote", name, int(box["width"]), "x", int(box["height"]))
        print("errors:", errs or "none")
    b.close()

File: preview/shoot.mjs
// Screenshot helper (Playwright + the pre-installed Chromium).
//   node preview/shoot.mjs page index.html preview/desktop-full.png 1440 900 full
//   node preview/shoot.mjs page index.html preview/mobile-full.png 390 844 full mobile
//   node preview/shoot.mjs banners banners/banners.html banners
// Prints loaded fonts, horizontal-overflow check, page height and any console errors.
import { pathToFileURL } from 'node:url';
import { resolve } from 'node:path';
import { mkdirSync } from 'node:fs';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
let chromium;
try { ({ chromium } = require('playwright')); }
catch { ({ chromium } = require('/opt/node22/lib/node_modules/playwright')); }

const [mode, src, out, w, h, full, mobile] = process.argv.slice(2);
const launch = { executablePath: process.env.CHROME_PATH || undefined };
const browser = await chromium.launch(launch);

// Optional: serve Google Fonts from a local cache (FONT_DIR holds local.css + f*.woff2 fetched with curl),
// for environments where the headless browser cannot reach fonts.googleapis.com directly.
async function routeFonts(ctx) {
  const dir = process.env.FONT_DIR;
  if (!dir) return;
  const { readFileSync, existsSync } = await import('node:fs');
  await ctx.route(/https:\/\/fonts\.googleapis\.com\/.*/, r => r.fulfill({ status: 200, contentType: 'text/css', body: readFileSync(`${dir}/local.css`, 'utf8') }));
  await ctx.route(/https:\/\/fonts\.gstatic\.com\/local\/(.+)/, r => {
    const f = `${dir}/${r.request().url().split('/local/')[1]}`;
    return existsSync(f) ? r.fulfill({ status: 200, contentType: 'font/woff2', body: readFileSync(f) }) : r.abort();
  });
}

if (mode === 'page') {
  const isMobile = mobile === 'mobile';
  const ctx = await browser.newContext({ viewport: { width: +w, height: +h }, deviceScaleFactor: isMobile ? 2 : 1, isMobile, hasTouch: isMobile });
  await routeFonts(ctx);
  const pg = await ctx.newPage();
  const errs = [];
  pg.on('pageerror', e => errs.push(String(e)));
  pg.on('console', m => { if (m.type() === 'error' || m.type() === 'warning') errs.push(`console.${m.type()}: ${m.text()}`); });
  await pg.goto(pathToFileURL(resolve(src)).href);
  await pg.evaluate(() => document.fonts.ready);
  await pg.waitForTimeout(1200);
  // force pending reveals + rails so a full-page capture is complete
  await pg.evaluate(() => {
    document.querySelectorAll('.reveal-pending').forEach(e => e.classList.add('in'));
    document.querySelectorAll('[data-rail-item]').forEach(e => e.classList.add('lit'));
    document.querySelectorAll('.rail').forEach(e => e.style.setProperty('--progress', 1));
    document.querySelector('#grow')?.classList.add('in');
  });
  await pg.waitForTimeout(full === 'full' ? 4200 : 600);
  const report = await pg.evaluate(() => ({
    fonts: [...document.fonts].filter(f => f.status === 'loaded').map(f => f.family).filter((v, i, a) => a.indexOf(v) === i),
    overflow: document.documentElement.scrollWidth > document.documentElement.clientWidth,
    scrollW: document.documentElement.scrollWidth, clientW: document.documentElement.clientWidth,
    h: document.documentElement.scrollHeight
  }));
  console.log('fonts:', report.fonts.join(', '));
  console.log('overflow:', report.overflow, report.scrollW, report.clientW, 'height:', report.h);
  console.log('errors:', errs.length ? errs : 'none');
  await pg.screenshot({ path: out, fullPage: full === 'full' });
  console.log('wrote', out);
} else if (mode === 'banners') {
  mkdirSync(out, { recursive: true });
  const ctx = await browser.newContext({ viewport: { width: 2600, height: 2000 }, deviceScaleFactor: 1 });
  await routeFonts(ctx);
  const pg = await ctx.newPage();
  const errs = [];
  pg.on('pageerror', e => errs.push(String(e)));
  await pg.goto(pathToFileURL(resolve(src)).href);
  await pg.evaluate(() => document.fonts.ready);
  await pg.waitForTimeout(1500);
  for (const el of await pg.$$('.canvas')) {
    const name = await el.getAttribute('data-name');
    const box = await el.boundingBox();
    await el.screenshot({ path: `${out}/${name}.png` });
    console.log('wrote', name, Math.round(box.width), 'x', Math.round(box.height));
  }
  console.log('errors:', errs.length ? errs : 'none');
}
await browser.close();
